src/keyboard/plugins/character.ts

```diff
@@ -147,12 +147,65 @@ export const keypressBehavior: behaviorPlugin[] = [
         },
       })
 
-      const appliedValue = getValue(element)
-      if (appliedValue === newValue) {
-        state.carryValue = undefined
-      } else {
-        state.carryValue = newValue
-      }
+      setCarryValue(element, state, newValue)
+    },
+  },
+  {
+    matches: (keyDef, element) =>
+      keyDef.key?.length === 1 &&
+      isElementType(element, 'input', {type: 'email', readOnly: false}),
+    handle: (keyDef, element, options, state) => {
+      const oldValue =
+        state.carryValue ?? getValue(element) ?? /* istanbul ignore next */ ''
+
+      const {newValue, newSelectionStart} = calculateNewValue(
+        keyDef.key as string,
+        element as HTMLElement,
+        oldValue,
+      )
+
+      fireInputEvent(element as HTMLElement, {
+        newValue,
+        newSelectionStart,
+        eventOverrides: {
+          data: keyDef.key,
+          inputType: 'insertText',
+        },
+      })
+
+      setCarryValue(element, state, newValue)
+    },
+  },
+  {
+    matches: (keyDef, element) =>
+      keyDef.key?.length === 1 && isElementType(element, 'input', {type: 'url', readOnly: false}
+      ),
+    handle: (keyDef, element, options, state) => {
+      const oldValue =
+        state.carryValue ?? getValue(element) ?? /* istanbul ignore next */ ''
+
+      const selectionRange = oldValue.endsWith(" ") ? {
```

Comment:
I think this should check for `state.carryValue` being `undefined`.
Otherwise this adds the implicit and wrong assumption that `Space` at the end of the value is the only possible case for `carryValue`.

Reply:
That's a very good point, I'll add a undefined check in